app/utils/validation.ts

```diff
@@ -109,23 +113,59 @@ export const ifNotField =
     context[field] ? ([true] as const) : validator(value, context);
 
 export function createValidator(
-  fieldValidators: { [field: string]: ReturnType<Validator>[] },
-  rawValidator?: (input) => ValidatorResult
+  fieldValidators: {
+    [field: string]: (ReturnType<Validator> | ReturnType<AsyncValidator>)[];
+  },
+  rawValidator?: (input) => ValidatorResult,
+  makeAsync = false
 ) {
-  return function validate(input) {
-    const rawValidatorErrors: ValidatorResult = rawValidator?.(input) || {};
-    const fieldValidatorErrors = Object.keys(fieldValidators).reduce(
-      (errors: ValidatorResult, field) => {
-        const fieldErrors =
-          fieldValidators[field]
-            .map((validator) => validator(input[field], input))
-            .filter(([isValid]) => !isValid)
-            .map(([, message]) => message || 'not valid') || 0;
-        if (fieldErrors.length) errors[field] = fieldErrors;
-        return errors;
-      },
-      {}
-    );
-    return merge(fieldValidatorErrors, rawValidatorErrors);
-  };
+  return makeAsync
+    ? async function validate(input) {
+        const rawValidatorErrors: ValidatorResult = rawValidator?.(input) || {};
+
+        const validationResults = {};
+
+        for (let field of Object.keys(fieldValidators)) {
+          validationResults[field] = await Promise.all(
+            fieldValidators[field].map(async (validator) => {
+              let result = validator(input[field], input);
+              if (result instanceof Promise) {
+                result = await result;
+              }
+              return result;
+            })
+          );
+        }
+
+        const fieldValidatorErrors = Object.keys(fieldValidators).reduce(
+          (errors: ValidatorResult, field) => {
+            const fieldErrors =
+              validationResults[field]
+                .filter(([isValid]) => !isValid)
+                .map(([, message]) => message || 'not valid') || 0;
```

Comment:
The `|| 0` part seems out of place. If the intention was to provide a default value, I believe this won't work as expected because the map function will always return an array, making the `|| 0` part redundant.

Reply:
Agreed. It is also technically unreachable the way it is written now but I've refactored it into something that makes more sense.